### PR TITLE
3.next email transport

### PR DESCRIPTION
**Pull request by ADmad**
Added `TransportFactory` and `TransportRegistry` to off load transport instance creation responsibilities from the `Email` class and deprecated related methods.

### CODE REVIEW

File: src/Mailer/TransportFactory.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            );
+        }
+
+        if (is_array(static::$_config[$name]) && empty(static::$_config[$name]['className'])) {
+            throw new InvalidArgumentException(
+                sprintf('Transport config "%s" is invalid, the required `className` option is missing', $name)
+            );
+        }
+
+        static::getRegistry()->load($name, static::$_config[$name]);
```
Comment (markstory) [inline src/Mailer/TransportFactory.php]: This method looks like what is sometimes found in the `_create` method of a registry implementation. What does having a factory + registry class provide us?

File: src/Mailer/TransportRegistry.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        if (is_object($class)) {
+            return $class;
+        }
+
+        $className = App::className($class, 'Mailer/Transport', 'Transport');
+
+        if (!$className) {
+            $className = App::className($class, 'Network/Email', 'Transport');
+            if ($className) {
+                trigger_error(
```
Comment (markstory) [inline src/Mailer/TransportRegistry.php]: Should this use `deprecationWarning()` now? I think this codes previous home predates that function's existence.

**Review by markstory — COMMENTED** [review]
I like the separation of concern out of `Email`.

File: src/Mailer/TransportRegistry.php (hunk 2 above)
Comment (ADmad) [reply src/Mailer/TransportRegistry.php]: Yes it should, I'll update.

File: src/Mailer/TransportFactory.php (hunk 1 above)
Comment (markstory) [reply src/Mailer/TransportFactory.php]: What if we only had the registry? A registry in many ways is a memoized factory.

Comment (ADmad) [reply src/Mailer/TransportFactory.php]: If we only had the registry then we would have to make the registry use the `StaticConfigTrait` to provide the `setConfg()`, `getConfig()` etc. static methods.

Currently I have made the `TranportFactory`-`TranportRegistry` pair similar to `Cache`- `CacheRegistry` and `Log`-`LogEngineRegistry` pairs we have (without the extra methods `Cache` and `Log` have).

Comment (markstory) [reply src/Mailer/TransportFactory.php]: Ok. Lets roll with that until the DI apocalypse arrives :smile:

Comment (markstory) [reply src/Mailer/TransportFactory.php]: Lets keep this.

**Comment by ADmad** [comment]
This is good to go now.

**Pull request merged**